### PR TITLE
[HLSL][DXIL] Implement WaveGetLaneIndex Intrinsic

### LINKED ISSUE #70105: [HLSL] implement `WaveGetLaneIndex` intrinsic
Implement HLSL `WaveGetLaneIndex` intrinsic:

https://learn.microsoft.com/en-us/windows/win32/direct3dhlsl/WaveGetLaneIndex

AC:
* The intrinsic is implemented and lowered to LLVM-IR
* Tests are implemented to validate that the IR is generated correctly

### PR DESCRIPTION
**Pull request by inbelic**
- add additional lowering for directx backend in CGBuiltin.cpp
    - add directx intrinsic to IntrinsicsDirectX.td
    - add semantic check of arguments in SemaHLSL.cpp
    - add mapping to DXIL op in DXIL.td

    - add testing of semantics in WaveGetLaneIndex-errors.hlsl
    - add testing of dxil lowering in WaveGetLaneIndex.ll
  
Resolves #70105

### CODE REVIEW

File: clang/lib/CodeGen/CGBuiltin.cpp [hunk 1]
```diff
@@ -18827,9 +18827,21 @@ case Builtin::BI__builtin_hlsl_elementwise_isinf: {
         ArrayRef<Value *>{Op0, Op1}, nullptr, "hlsl.step");
   }
   case Builtin::BI__builtin_hlsl_wave_get_lane_index: {
-    return EmitRuntimeCall(CGM.CreateRuntimeFunction(
-        llvm::FunctionType::get(IntTy, {}, false), "__hlsl_wave_get_lane_index",
-        {}, false, true));
+    // Since we don't define a SPIR-V intrinsic for the SPIR-V built-in from
+    // SPIRVBuiltins.td, manually get the matching name for the DirectX
+    // intrinsic and the demangled builtin name
```
Comment (pow2clk) [inline clang/lib/CodeGen/CGBuiltin.cpp:18832]: I fear I missed some discussion explaining this, but why don't we add a builtin to SPIRVBuiltins.td?

File: clang/test/SemaHLSL/BuiltIns/WaveGetLaneIndex-errors.hlsl [hunk 2]
```diff
@@ -0,0 +1,6 @@
+// RUN: %clang_cc1 -finclude-default-header -triple dxil-pc-shadermodel6.6-library %s -fnative-half-type -emit-llvm-only -disable-llvm-passes -verify -verify-ignore-unexpected
```
Comment (pow2clk) [inline clang/test/SemaHLSL/BuiltIns/WaveGetLaneIndex-errors.hlsl:1]: Are you getting unexpected diagnostics? If so, I'd like to understand them. If not, I don't think `-verify-ignore-unexpected` is needed. If anything, we should limit it to the type of diagnostic that is being produced e.g. `-verify-ignore-unexpected=note`

**Review by pow2clk — COMMENTED** [review]
Small typo in description: 
"IntrinscsDirectX"

Looks good, but I have a question about the SPIRV part.

File: clang/test/SemaHLSL/BuiltIns/WaveGetLaneIndex-errors.hlsl (hunk 2 above)
Comment (inbelic) [reply clang/test/SemaHLSL/BuiltIns/WaveGetLaneIndex-errors.hlsl]: The flag was not needed, so have removed it.

**Review by farzonl — APPROVED** [review]

**Review by pow2clk — APPROVED** [review]
Looks good! I since remembered that I can edit descriptions and fix small typos like that in the future 🙃

**Pull request merged**